### PR TITLE
[FIX] Message_AllowedMaxSize fails for emoji sequences

### LINKED ISSUE #10422: [BUG] "Message_AllowedMaxSize" not working as expected with messages containing emoji unicode sequences
If "Message_AllowedMaxSize" = 10

This message will be rejected: "123456789👨‍👨‍👧‍👦"

### PR DESCRIPTION
**Pull request by c0dzilla**
Closes #10422 

This pr ensures emojis are treated as single characters when determining if message size exceeds allowed limit, for better experience.

Taking same example as mentioned in issue, if max allowed size is 10
image

now works and has effective size 10.

### CODE REVIEW

**Comment by cardoso** [comment]
Is this working only for :shortname: ? 

Or is it covering also the case where the user types the actual "👨‍👨‍👧‍👦" emoji instead of its :shortname: . From the iOS emoji keyboard for instance.

When I said emoji sequences I meant unicode sequences. And Javascript can treat them as multiple characters unfortunately.

**Comment by c0dzilla** [comment]
@sampaiodiego I've converted emoji unicodes to shortnames using `toShort()`. I think it should handle that as well now.

**Comment by cardoso** [comment]
@c0dzilla @sampaiodiego we need a better solution than this. Converting back to shortname would imply that this solution only works for unicode graphemes with a corresponding shortname, which is very limiting. The iOS App for instance has many more emojis than in the Web App.

We need a generic solution for counting unicode graphemes. From what I've searched, it's non-trivial in JS and we may need a new dependency, but I think it's worth it.

File: packages/rocketchat-lib/server/methods/sendMessage.js [hunk 1]
```diff
@@ -29,10 +31,22 @@ Meteor.methods({
...
 		}
 
-		if (message.msg && message.msg.length > RocketChat.settings.get('Message_MaxAllowedSize')) {
-			throw new Meteor.Error('error-message-size-exceeded', 'Message size exceeds Message_MaxAllowedSize', {
-				method: 'sendMessage'
+		if (message.msg) {
+			// converting emoji unicodes to shortnames if present
+			let adjustedMessage = emojione.toShort(message.msg);
+
+			adjustedMessage = adjustedMessage.replace(/:\w+:/gm, (match) => {
```
Comment (cardoso) [inline packages/rocketchat-lib/server/methods/sendMessage.js]: Why is this part still needed? 🤔

**Comment by graywolf336** [comment]
Why not use `Array.from` instead of underscore's?

**Comment by cardoso** [comment]
Also, I think Unit Tests will be a good idea here 👍 

Testing cases like the one pointed out in the issue "123456789👨‍👨‍👧‍👦" == 10

**Comment by c0dzilla** [comment]
After having a look at https://news.ycombinator.com/item?id=13830177, it seems that converting string to array will not give the right length always(like for ❤️ it gives length 2 instead of 1).

**Comment by c0dzilla** [comment]
I've used (grapheme-splitter)[https://github.com/orling/grapheme-splitter] module, it solves the issue.

**Comment by c0dzilla** [comment]
@graywolf336 @rodrigok please have a look. @cardoso I'll add tests soon.

**Comment by c0dzilla** [comment]
@sampaiodiego @graywolf336 please have a look.

File: packages/rocketchat-ui/client/lib/chatMessages.js [hunk 2]
```diff
@@ -554,7 +554,13 @@ this.ChatMessages = class ChatMessages {
...
 
 	isMessageTooLong(message) {
-		return message && message.length > this.messageMaxSize;
+		const adjustedMessage = message.replace(/:\w+:/gm, (match) => {
+			if (RocketChat.emoji.list[match] !== undefined) {
+				return ' ';
+			}
+			return match;
+		});
+		return message && RocketChat.messageProperties(adjustedMessage).length > this.messageMaxSize;
```
Comment (ggazzo) [inline packages/rocketchat-ui/client/lib/chatMessages.js]: this test on 'message' should not be done before?

File: packages/rocketchat-ui/client/lib/chatMessages.js [hunk 3]
```diff
@@ -554,7 +554,13 @@ this.ChatMessages = class ChatMessages {
 	}
 
 	isMessageTooLong(message) {
-		return message && message.length > this.messageMaxSize;
+		const adjustedMessage = message.replace(/:\w+:/gm, (match) => {
```
Comment (ggazzo) [inline packages/rocketchat-ui/client/lib/chatMessages.js]: I think we have this replace in two places... maybe use an export/import? (:

File: packages/rocketchat-ui/client/lib/chatMessages.js (hunk 2 above)
Comment (c0dzilla) [reply packages/rocketchat-ui/client/lib/chatMessages.js]: Yeah, sorry about that :sweat_smile:

**Comment by c0dzilla** [comment]
@ggazzo I've made the changes :relaxed:

**Pull request merged**